### PR TITLE
Student attendance tracking system (v2 Feature Branch)

### LINKED ISSUE #17: [FEAT] Implement Student Attendance Tracking System with ERP Integration
### **Description**

Develop a tracking system that records student check-in and check-out times at school. The system should automatically identify late arrivals and absences and synchronize all attendance data with the school ERP system.

**Key requirements:**

* Track student entry and exit times
* Detect and mark lateness and absences
* Store attendance records reliably
* Synchronize attendance data with the school ERP
* Ensure data consistency and accuracy across systems

### LINKED ISSUE #26: [v2] Add /admin-api/reports/visits endpoint for paginated visitor attendance reports
Implement a new reporting endpoint that returns paginated visitor attendance information aggregated per visitor per day for a given date range. The endpoint should support optional filters and return sign-in / sign-out timestamps, counts and duration based on the provided SQL logic.

⸻

Endpoint

`GET /admin-api/reports/visits`

⸻

Request parameters

**Required**
	•	`from` (date, YYYY-MM-DD) — start date (inclusive). Minimal precision: day.
	•	`to` (date, YYYY-MM-DD) — end date (inclusive).

_Optional_
	•	`is_student` (boolean) — filter by student flag.
	•	`year_group` (integer) — filter by year group.
	•	`sign_status` (string) — filter by sign status (not_signed, signed_in, signed_out).
	•	`page` (integer, default 1) — page number.
	•	`limit` (integer, default 20) — records per page.

Date range parameters are day-precise. Convert to to end-of-day in queries (e.g. to + 1 day exclusive) or use inclusive date logic.

Response format (JSON)

```
{
  "page": 1,
  "limit": 20,
  "total": 154,
  "data": [
    {
      "visit_date": "2026-03-10",
      "visitor_id": 25,
      "name": "John",
      "surname": "Smith",
      "is_student": true,
      "year_group": 10,
      "visits_count": 2,
      "sign_status": "signed_out",
      "signed_in": "2026-03-10T08:45:12Z",
      "signed_out": "2026-03-10T15:12:55Z",
      "duration_minutes": 387
    },
    {
      "visit_date": "2026-03-10",
      "visitor_id": 31,
      "name": "Anna",
      "surname": "Brown",
      "is_student": true,
      "year_group": 9,
      "visits_count": 1,
      "sign_status": "signed_in",
      "signed_in": "2026-03-10T09:02:33Z",
      "signed_out": null,
      "duration_minutes": null
    }
  ]
}

```

⸻

Sign status logic
	•	`not_signed` — visits_count = 0
	•	`signed_in` — visits_count is odd (1,3,…) — visitor currently signed in
	•	`signed_out` — visits_count is even (2,4,…) — visitor signed out

⸻

Acceptance criteria
	•	Endpoint GET /admin-api/reports/visits implemented and documented.
	•	from and to date parameters are required and interpreted at day granularity.
	•	Results are paginated (page, limit) and include total count.
	•	Filters is_student, year_group, sign_status apply correctly.
	•	Response fields match the defined response format (including signed_in, signed_out, duration_minutes).
	•	Sign-in/out timestamps and duration_minutes calculated per the SQL logic.
	•	Endpoint performs efficiently for large datasets (no full-table cross-joins without justification).
	•	Unit/integration tests cover core logic and edge cases (no visits, single visit, multiple visits, filters).
	•	API documentation (OpenAPI/Swagger) updated.

⸻

Implementation notes / suggestions
	•	Convert from/to into [start_datetime, end_datetime) for SQL (e.g. from 00:00:00 to to + 1 day 00:00:00 exclusive).
	•	Avoid CROSS JOIN visitors for production if the visitors table is large — instead: generate the date series and join only visitors who have tracks in the period, or pre-aggregate tracker.track by date and visitor then join to visitors.
	•	Add indexes on tracker.track(visitor_id, created_at) and on visitors(id) if not present.
	•	Consider adding csv/excel export option for reports (future extension).
	•	Return timestamps in ISO-8601 UTC (or include timezone explicitly).

⸻

If you want, I can also generate an OpenAPI snippet (YAML) and a basic controller skeleton you can paste into the repo.

### PR DESCRIPTION
**Pull request by buzyka**
closes #17 
closes #26

### CODE REVIEW

File: internal/usecase/synchroniser/isams_sync.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		fullName = *student.FullName
+	}
+
+	var grade, yearGroup int
+	if student.YearGroup != nil {
+		grade = *student.YearGroup
+		yearGroup = *student.YearGroup
+	}
+
+	devisions, err := s.getDivisionsByYearGroup(int32(yearGroup))
```
[the remark this answers is not in the document]
Comment (buzyka) [reply internal/usecase/synchroniser/isams_sync.go]: fixed

File: internal/infrastructure/repository/visitor_sqlite.go [hunk 2]
```diff
@@ -44,25 +105,48 @@ func (r *Visitor) FindByKey(key string) (*entity.VisitDetails, error) {
...
 	var tmpGrade sql.NullInt32
 	var tmpImage sql.NullString
-	
-	row := r.Connection.QueryRow("SELECT id, name, surname, grade, image FROM visitors WHERE id = ?", id)
+	var tmpErpID sql.NullInt64
+	var tmpErpSchoolID sql.NullString
+	var tmpYearGroup sql.NullInt32
+	var tmpDivisions sql.NullString
+
+	row := r.Connection.QueryRow("SELECT id, name, surname, grade, image, isams_id, isams_school_id, year_group, divisions FROM visitors WHERE id = ?", id)
```
[the remark this answers is not in the document]
Comment (buzyka) [reply internal/infrastructure/repository/visitor_sqlite.go]: fixed

File: internal/domain/provider/visitor_repository.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+type VisitorRepository interface {	
+	GetAll() ([]*entity.Visitor, error)
+	FindById(id int32) (*entity.Visitor, error)
+	FindByKey(key string) (*entity.VisitDetails, error)
+	
+	AddKeyToVisitor(visitor *entity.Visitor, key string) error
+	AddVisitor(visitor *entity.Visitor) error
+
+	
+}
```
[the remark this answers is not in the document]
Comment (buzyka) [reply internal/domain/provider/visitor_repository.go]: fixed

File: internal/infrastructure/gocontainer/gocontainer.go [hunk 4]
```diff
@@ -42,4 +54,19 @@ func Build(cfg *config.Config) {
...
+		}
+		return &ERPFactory{f: f}
+	})
+
+	container.MustSingleton(container.Global, func () provider.VisitorRepository {
+		return &repository.Visitor{
+			Connection: connection,
+		}
+	})
+}
```
[the remark this answers is not in the document]
Comment (buzyka) [reply internal/infrastructure/gocontainer/gocontainer.go]: fixed

File: internal/usecase/synchroniser/isams_sync_test.go [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+}
+
+func TestSaveStudent_IsUpToDate_UpdateNeeded(t *testing.T) {
+	mockRepo := new(MockVisitorRepository)
+	mockClient := new(MockERPClient)
+
+	sync := &StudentSync{
+		VisitorRepo: mockRepo,
+		currentClient: mockClient,
+		yearGroupDevisions: make(map[int32][]int32),
```
[the remark this answers is not in the document]
Comment (buzyka) [reply internal/usecase/synchroniser/isams_sync_test.go]: fixed

File: internal/infrastructure/repository/visitor_sqlite.go [hunk 6]
```diff
@@ -14,24 +15,84 @@ type Visitor struct {
...
-	
+	var tmpErpID sql.NullInt64
+	var tmpErpSchoolID sql.NullString
+	var tmpYearGroup sql.NullInt32
+	var tmpDivisions sql.NullString
+
 	key = strings.ToUpper(key)
-	row := r.Connection.QueryRow("SELECT v.id, v.name, v.surname, v.grade, v.image, vk.key_id FROM visitors AS v INNER JOIN visitor_key AS vk ON vk.visitor_id = v.id WHERE vk.key_id = ?", key)
-	
+	row := r.Connection.QueryRow("SELECT v.id, v.name, v.surname, v.grade, v.image, v.isams_id, v.isams_school_id, v.year_group, v.divisions, vk.key_id FROM visitors AS v INNER JOIN visitor_key AS vk ON vk.visitor_id = v.id WHERE vk.key_id = ?", key)
```
[the remark this answers is not in the document]
Comment (buzyka) [reply internal/infrastructure/repository/visitor_sqlite.go]: fixed

File: internal/infrastructure/integration/isams/client.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+type Client struct {
+	BaseURL    string
+	HTTPClient *http.Client
+}
+
+
+func (c *Client) Do(req *http.Request) (*http.Response, error) {
+	if resp, err :=httputil.DumpRequestOut(req, true); err == nil {
+		fmt.Printf("Request: %s\n", string(resp))
+	}
```
[the remark this answers is not in the document]
Comment (buzyka) [reply internal/infrastructure/integration/isams/client.go]: fixed

File: internal/infrastructure/repository/visitor_sqlite.go [hunk 8]
```diff
@@ -101,9 +200,115 @@ func (r *Visitor) AddKeyToVisitor(visitor *entity.Visitor, key string) error {
...
+		string(divisionsStr),
+		visitor.UpdatedAt,
+		fmt.Sprintf("%d", visitor.GetSyncHash()),
+	)
+	if err != nil {
+		return err
+	}
+	id, err := result.LastInsertId()
+	if err != nil {
+		return fmt.Errorf("visitor has not be saved correctly: %s", err.Error())
```
[the remark this answers is not in the document]
Comment (buzyka) [reply internal/infrastructure/repository/visitor_sqlite.go]: fixed

File: internal/domain/provider/visitor_repository.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (buzyka) [reply internal/domain/provider/visitor_repository.go:34]: fixed

File: internal/isb/entity/visitor.go [hunk 9]
```diff
@@ -1,14 +1,59 @@
...
+	h.Write([]byte{0x00})
+
+	// Write ErpYearGroupID (int32)
+	_ = binary.Write(h, binary.BigEndian, v.ErpYearGroupID)
+	h.Write([]byte{0x00})
+
+	// Write ErpDivisions ([]int32)
+	for _, div := range divs {
+		_ = binary.Write(h, binary.BigEndian, div)
+	}
```
[the remark this answers is not in the document]
Comment (buzyka) [reply internal/isb/entity/visitor.go]: leave it like this so far

**Comment by buzyka** [comment]
merged in  https://github.com/buzyka/imlate/pull/21
